core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -528,24 +531,94 @@ public class RexImpTable {
   public static final MemberExpression BOXED_TRUE_EXPR =
       Expressions.field(null, Boolean.class, "TRUE");
 
-  private final ImmutableMap<SqlOperator, RexCallImplementor> map;
+  private final ImmutableMultimap<SqlOperator, OpImplementatorPair> scalarMap;
   private final ImmutableMap<SqlAggFunction, Supplier<? extends AggImplementor>> aggMap;
   private final ImmutableMap<SqlAggFunction, Supplier<? extends WinAggImplementor>> winAggMap;
   private final ImmutableMap<SqlMatchFunction, Supplier<? extends MatchImplementor>> matchMap;
   private final ImmutableMap<SqlOperator, Supplier<? extends TableFunctionCallImplementor>>
       tvfImplementorMap;
 
   private RexImpTable(Builder builder) {
-    this.map = ImmutableMap.copyOf(builder.map);
+    this.scalarMap = ImmutableMultimap.copyOf(builder.scalarMap.map);
     this.aggMap = ImmutableMap.copyOf(builder.aggMap);
     this.winAggMap = ImmutableMap.copyOf(builder.winAggMap);
     this.matchMap = ImmutableMap.copyOf(builder.matchMap);
     this.tvfImplementorMap = ImmutableMap.copyOf(builder.tvfImplementorMap);
   }
 
+  /**
+   * A pairing of an SqlOperator and a RexCallImplementor. OpImplementor pairs
+   * will be values in the scalar operators map. This allows an "==" check on
+   * the SqlOperator to select the correct operator since the hash of operators
+   * may not be unique.
+   */
+  private static class OpImplementatorPair {
```

Comment:
Could you use `Pair<SqlOperator, RexCallImplementor>` instead? I don't think `hashCode` or `equals` are ever called on the values in a map or array-list-multimap.

Reply:
Or, maybe better, change `scalarMap` to type `Map<SqlOperator, ImmutablePairList<SqlOperator, RexCallImplementor>>`.

Reply:
I have rebased this on top of your change to refactor `RexImpTable`. I am now using `ImmutablePairList`.